src/Stratis.Bitcoin.Features.Miner/PowMining.cs

```diff
@@ -168,110 +168,181 @@ public void StopMining()
+            context.ChainTip = this.consensusLoop.Tip;
+            if (this.chain.Tip != context.ChainTip)
+            {
+                Task.Delay(TimeSpan.FromMinutes(1), this.nodeLifetime.ApplicationStopping).GetAwaiter().GetResult();
+                return false;
+            }
 
-                if (blockValidationContext.ChainedHeader == null)
-                {
-                    this.logger.LogTrace("(-)[REORG-2]");
-                    return blocks;
-                }
+            return true;
+        }
 
-                if (blockValidationContext.Error != null)
-                {
-                    if (blockValidationContext.Error == ConsensusErrors.InvalidPrevTip)
-                        continue;
+        private bool BuildBlock(MineBlockContext context)
+        {
+            context.BlockTemplate = this.blockProvider.BuildPowBlock(context.ChainTip, context.ReserveScript.ReserveFullNodeScript);
 
-                    this.logger.LogTrace("(-)[ACCEPT_BLOCK_ERROR]");
-                    return blocks;
-                }
+            if (this.network.Consensus.IsProofOfStake)
+            {
+                // Make sure the POS consensus rules are valid. This is required for generation of blocks inside tests,
+                // where it is possible to generate multiple blocks within one second.
+                if (context.BlockTemplate.Block.Header.Time <= context.ChainTip.Header.Time)
+                    return false;
+            }
+
+            return true;
+        }
+
+        private bool MineBlock(MineBlockContext context)
+        {
+            context.ExtraNonce = this.IncrementExtraNonce(context.BlockTemplate.Block, context.ChainTip, context.ExtraNonce);
+
+            Block block = context.BlockTemplate.Block;
+            while ((context.MaxTries > 0) && (block.Header.Nonce < InnerLoopCount) && !block.CheckProofOfWork())
+            {
+                this.miningCancellationTokenSource.Token.ThrowIfCancellationRequested();
+
+                ++block.Header.Nonce;
+                --context.MaxTries;
+            }
+
+            if (context.MaxTries == 0)
+                return false;
+
+            return true;
+        }
+
+        private bool ValidateMinedBlock(MineBlockContext context)
+        {
+            if (context.BlockTemplate.Block.Header.Nonce == InnerLoopCount)
+                return false;
+
+            var chainedHeader = new ChainedHeader(context.BlockTemplate.Block.Header, context.BlockTemplate.Block.GetHash(), context.ChainTip);
+            if (chainedHeader.ChainWork <= context.ChainTip.ChainWork)
+                return false;
+
+            return true;
+        }
 
-                this.logger.LogInformation("Mined new {0} block: '{1}'.", BlockStake.IsProofOfStake(blockValidationContext.Block) ? "POS" : "POW", blockValidationContext.ChainedHeader);
+        private void ValidateMinedBlockWithConsensus(MineBlockContext context)
```

Comment:
I would suggest to merge this two methods `ValidateMinedBlockWithConsensus` `CheckValidationContext`
And call it `ValidateAndconnectBlock()`

Reply:
💯